src/System.Security.Cryptography.Cng/tests/ImportExportTests.cs

```diff
@@ -46,5 +45,191 @@ public static void TestImportExportRoundTrip()
+            {
+                param = cng.ExportExplicitCurveParameters(false);
+                Assert.True(param.Curve is ECPrimeCurve);
+                using (newCng = new ECDsaCng())
+                {
+                    newCng.ImportParameters(param);
+                    param2 = newCng.ExportExplicitCurveParameters(false);
+                    IsEqual(param, param2);
+                }
+            }
+        }
+
+        [Fact]
+        public static void TestRegenKeyNegative()
+        {
+            if (!EcDSACngTests.IsEccNamedCurveSupported())
+                return;
+
+            ECParameters param;
+            ECDsaCng cng, newCng;
+
+            using (cng = new ECDsaCng(ECNamedCurve.brainpoolP160r1))
+            {
+                param = cng.ExportExplicitCurveParameters(true);
+                Assert.True(param.Curve is ECPrimeCurve);
+                Assert.NotEqual(param.D, null);
+                using (newCng = new ECDsaCng())
+                {
+                    newCng.ImportParameters(param);
+
+                    // the curve name is not flowed on on explicit export\import (by design) so we are unable to generate new
+                    Assert.ThrowsAny<CryptographicException>(() => newCng.GenerateNewKey());
+                }
+            }
+        }
+
+        [Fact]
+        public static void TestRegenKeyPositive()
+        {
+            if (!EcDSACngTests.IsEccNamedCurveSupported())
+                return;
+
+            ECParameters param, param2;
+            ECDsaCng cng;
+
+            using (cng = new ECDsaCng(ECNamedCurve.brainpoolP160r1))
+            {
+                param = cng.ExportExplicitCurveParameters(true);
+                Assert.True(param.Curve is ECPrimeCurve);
+                Assert.NotEqual(param.D, null);
+
+                cng.GenerateNewKey();
+                param2 = cng.ExportExplicitCurveParameters(true);
+
+                ComparePublicKey(param.Q, param2.Q, false); // public key should not match
+                Assert.False(IsEqual(param.D, param2.D)); // private key should not match
+                CompareCurve(param.Curve, param2.Curve); // curve should match
+            }
+        }
+
+        private static void IsEqual(ECParameters p1, ECParameters p2)
+        {
+            ComparePublicKey(p1.Q, p2.Q);
+            CompareCurve(p1.Curve, p2.Curve);
+        }
+
+        private static void ComparePublicKey(ECPoint q1, ECPoint q2, bool isEqual = true)
+        {
+            Assert.Equal(isEqual, IsEqual(q1.X, q2.X));
+            Assert.Equal(isEqual, IsEqual(q1.Y, q2.Y));
+        }
+
+        private static void CompareCurve(ECCurve c1, ECCurve c2)
+        {
+            Assert.True((c1 == null && c2 == null) || (c1 != null && c2 != null));
+            if (c1 is ECNamedCurve)
+            {
+                ECNamedCurve curve1 = (ECNamedCurve)c1;
+                ECNamedCurve curve2 = (ECNamedCurve)c1;
```

Comment:
A) this should be `c2` :smile:  B) Probably want to Assert c2's type before casting; just for a cleaner error report.